libmusig2.c

```diff
@@ -1,282 +1,359 @@
+        signer->R[j] = (unsigned char*)malloc(PK_SIZE);
+
+		while (1) {
+            /* Generate a random r[j]. */
+			if (!fill_random(signer->r[j], SK_SIZE)) {
+                printf("Failed to generate randomness!\n");
+                return;
+			}
+            /* Generate the corresponding nonce R and serialize it in R[j]. */
+            if (secp256k1_ec_pubkey_create(ctx, pubkey, signer->r[j]) &&
+                    secp256k1_ec_pubkey_serialize(ctx, signer->R[j], &point_ser_len, pubkey, SECP256K1_EC_COMPRESSED))
+                break;
+		}
+    }
 }
 
-// Function that generates randomness and their corresponding commitments, for `batch_size` signatures.
-void batch_commit(unsigned char *commitment, unsigned char *randomness, unsigned long batch_size) {
-    for (int j = 0; j < batch_size; j ++){
-        for (int i = 0; i < NR_V; i++) {
-            crypto_core_ristretto255_scalar_random(randomness + j * NR_V * crypto_core_ristretto255_SCALARBYTES + i * crypto_core_ristretto255_SCALARBYTES);
+void COLLECT_Signers(AGG_ROUND1 aggRound1, const unsigned char* pk, unsigned char** R, int i)
+{
+    int k, j;
+    aggRound1->PK_LIST[i] = (unsigned char*)malloc(PK_SIZE);
 
-            crypto_scalarmult_ristretto255_base(
-                    commitment + j * NR_V * crypto_core_ristretto255_SCALARBYTES + i * crypto_core_ristretto255_BYTES,
-                    randomness + j * NR_V * crypto_core_ristretto255_SCALARBYTES + i * crypto_core_ristretto255_SCALARBYTES);
-        }
+    /* Copy the content of signers public key into public key list. */
+    for (k = 0; k < PK_SIZE; k++)
+        aggRound1->PK_LIST[i][k] = pk[k];
+
+    for (j = 0;  j<V ;j++) {
+        aggRound1->R_LIST[j]->R[i] = (unsigned char*)malloc(PK_SIZE);
+
+        /* Copy the content of signers nonce into nonce list. */
+        for (k = 0; k < PK_SIZE; k++)
+            aggRound1->R_LIST[j]->R[i][k]= R[j][k];
     }
 }
 
-// Given a set of public keys, this function computes the exponent corresponding to
-// index `position`. This exponent is used when computing the aggregate signature.
-void compute_exponent(unsigned char *exponent, const unsigned char *pks, unsigned long position, unsigned long number_keys) {
-    // we use MAX (64) instead of normal (32) to get almost uniformity.
-    unsigned char hash[crypto_generichash_BYTES_MAX];
+void AGG_SIG_FirstRound(const secp256k1_context* ctx, AGG_ROUND1 aggRound1){
 
-    crypto_generichash_state state;
-    crypto_generichash_init(&state, NULL, 0, sizeof hash);
-    crypto_generichash_update(&state, pks, number_keys * crypto_core_ristretto255_BYTES);
-    crypto_generichash_update(&state, pks + position * crypto_core_ristretto255_BYTES, crypto_core_ristretto255_BYTES);
+    int i, j, k = 0;
+    size_t point_ser_len = PK_SIZE;
 
-    crypto_generichash_final(&state, hash, sizeof hash);
+    secp256k1_pubkey* temp_OUT = (secp256k1_pubkey*) malloc(sizeof (secp256k1_pubkey));
```

Comment:
does the secp256k1 API expose the ec point structure instead of the pubkey? essentially the same , but it might be less confusing.